docs/v3/core/index.rst

```diff
@@ -1599,46 +1471,266 @@ Storage transformers may be stacked to combine different functionalities:
+maintainer team reserves the right to refuse name assignment at its own
+discretion.
+
+- **Accepted regex:** ``^[a-z][a-z0-9-_.]+$``
+- **Valid examples:**
+  - ``zstd``
+  - ``numcodecs.adler32``
+- **Invalid examples:**
+  - ``foo/bar``
+
+.. _extension-naming-unregistered-names:
+
+Unregistered names
+^^^^^^^^^^^^^^^^^^
+
+Unregistered names are intended for private extensions and for experimental and development purposes.
+
+Unregistered names are not centrally managed and MAY be used by any extension without coordination.
+
+Unregistered names consist of URIs, which are prefixed with a scheme beginning
+with a letter and followed by any number of letters, numbers, plus symbols,
+dashes or dots and then followed by a colon.
+
+- **Identifying regex:** ``^([a-z][a-z0-9-_]+\.)+[a-z][a-z0-9-_]+:``
+
+URIs (`Uniform Resource Identifiers <https://en.wikipedia.org/wiki/Uniform_Resource_Identifier>`_)
+are a well-known mechanism to identify resources on the internet and extension authors are
+encouraged to explore further documentation on which identifiers might best express their intent.
+
+Aware that not all extension developers will want to immediately register a name,
+the goal of using URI identifiers is to provide a large and flexible namespace which
+balances the needs of developers building new extensions with a extensible mechanism
+which the Zarr community can make use of in the years to come.
+
+URLs (`Uniform Resource Locators <https://en.wikipedia.org/wiki/URL>`_) are one class of URIs which
+provide a mechanism for resolving resources.
+In previous versions of the v3 spec, the name of an extension was required to
+be a URI that dereferences to a human-readable codec specification, i.e. a URL.
+That is now discouraged for new extensions, though, for backwards compatibility
+with existing extensions, URLs names are still permitted.
+
+Instead, extension names SHOULD either be registered names as specified above or URNs.
+URNs (`Uniform Resource Names <https://en.wikipedia.org/wiki/Uniform_Resource_Name>`_)
+are simpler persistent identifiers assigned within defined namespaces.
+
+.. _extension-guidance:
+
+Guidance for extension authors
+------------------------------
+
+*This section is non-normative and provides assistance for the authors of
+extensions, especially those who are just getting started.*
+
+Recognizing that there are diverse considerations in choosing an extension
+name, guidance is provided below based on generic scenarios. Extension authors
+who are still unsure of how best to choose a name are welcome to open an issue
+on the zarr-specs repository.
+
+* **Local development**: Authors looking to define a name for local development
+  purposes should prefix their extensions with ``urn:x-``. This prefix defines
+  an "experimental" name. As such an extension matures, authors might consider
+  registering a new name for it. Implementations should check both for the
+  unregistered as well as the registered named.
```

Comment:
```suggestion
* **Local development of a new registered extension**: Authors developing a new extension intended for public
   consumption should feel free to coin a new non-prefixed "raw" name for use in development,
   with the understanding that the extension name should be registered by the time the extension is released.
```

Reply:
This suggestion feels correct, but it complicates the definition of "unregistered names", no? Isn't a name that has not yet been registered an unregistered name? Is the definition of "unregistered name" actually "a name which the creator has no intention of ever registering"? And that seems a bit hard to predict in advance.

Reply:
What do you suggest we do then?

Reply:
_If_ we stick with the meaning of "registered names" and "unregistered names" used earlier in this PR, we must require that local development uses a URI, regardless of the developer's intent to later register. Only when the developer registers a (non-URI) name can they use that name in their implementation

Because I think this conclusion is unacceptable, we have to revise the decision to have 1 type of name for registered extensions and another very different type of name for unregistered extensions. See https://github.com/zarr-developers/zarr-specs/pull/330#issuecomment-2759392714 for an alternative proposal.

Reply:
Resolving comment so we can merge the current simplified state of the PR.